jujugui/static/gui/src/app/components/machine-view/machine/machine.js

```diff
@@ -71,13 +73,140 @@ YUI.add('machine-view-machine', function() {
+    */
+    _updateConstraints: function(constraints) {
+      this.setState({constraints: constraints});
+    },
+
+    /**
+      Set the new constraints on the machine.
+    */
+    _setConstraints: function() {
+      const constraints = this.state.constraints;
+      const series = constraints.series || null;
+      delete constraints.series;
+      const id = this.props.machine.id;
+      this.props.updateMachineConstraints(id, constraints);
+      this.props.updateMachineSeries(id, series);
+      this._toggleForm();
+    },
+
+    /**
+      Generate the constraints form.
+
+      @returns {Object} the form JSX.
+    */
+    _generateConstraintsForm: function() {
+      if (!this.state.showForm) {
+        return null;
+      }
+      const machine = this.props.machine;
+      const disabled = this.props.acl.isReadOnly();
+      const units = this.props.units.filterByMachine(
+        machine.id, this.props.type === 'machine');
+      const buttons = [{
+        title: 'Cancel',
+        action: this._toggleForm,
+        type: 'base'
+      }, {
+        title: 'Update',
+        action: this._setConstraints,
+        type: 'neutral',
+        disabled: disabled
+      }];
+      return (
+        <div className="add-machine__constraints">
+          <h4 className="add-machine__title">
+            Update constraints
+          </h4>
+          <juju.components.Constraints
+            constraints={this._getConstraints()}
+            currentSeries={machine.series}
+            disabled={disabled}
+            hasUnit={!!units.length}
+            providerType={this.props.providerType}
+            series={this.props.series}
+            valuesChanged={this._updateConstraints} />
+          <juju.components.ButtonRow
+            buttons={buttons}
+            key="buttons" />
+        </div>);
+    },
+
+    /**
+      Get the constraints for the machine.
+
+      @returns {Object} the machine constraints.
+    */
+    _getConstraints: function() {
+      const constraints = this.props.machine.constraints || '';
+      let cpu;
+      let disk;
+      let mem;
+      let cpuCores;
+      let arch;
+      // The constraints are in the format:
+      // cpu-power=w cores=x mem=y root-disk=z
+      const parts = constraints.split(' ');
+      parts.forEach(part => {
+        const keyVal = part.split('=');
+        switch (keyVal[0]) {
+          case 'cpu-power':
+            cpu = keyVal[1];
+            break;
+          case 'cores':
+            cpuCores = keyVal[1];
+            break;
+          case 'mem':
+            mem = keyVal[1];
+            break;
+          case 'root-disk':
+            disk = keyVal[1];
+            break;
+          case 'arch':
+            arch = keyVal[1];
+            break;
+        }
+      });
+      return {
+        arch: arch || null,
+        cpuCores: cpuCores || null,
+        cpuPower: cpu || null,
+        disk: disk || null,
+        mem: mem || null
+      };
```

Comment:
Might be cleaner to do something like:
```
let types = {
  'cpu-power': null,
  cores: null,
  mem: null,
  'root-disk': null,
  arch: null;
};
```
So you can remove the switch and do
```
const keyVal = part.split('=');
types[keyVal[0]] = keyVal[1];
```
Then assign to what you want in the return
```
return {
  arch: types.arch,
  cpuCores: types.cores,
  cpuPower: types['cpu-power'],
  disk: types['root-disk'],
  mem: mem
};
```
Saves a few lines.
The ideal in this situation is to use a consistent naming convention, but that's not always possible.